### PR TITLE
Domaća zadaća 1 PR

### PR DESCRIPTION
**Pull request by JuraH32**

### CODE REVIEW

File: MovieApp/AppDelegate.swift [hunk 1]
```diff
@@ -6,6 +6,7 @@
 //
 
 import UIKit
+import PureLayout
```
Comment (IvanKulundzic) [inline MovieApp/AppDelegate.swift]: Why is this needed here?

File: MovieApp/CrewCollectionViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import UIKit
+import PureLayout
+
+class CrewCollectionViewCell: UICollectionViewCell {
+    let nameLabel = UILabel()
+    let roleLabel = UILabel()
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        self.commonInit()
```
Comment (IvanKulundzic) [inline MovieApp/CrewCollectionViewCell.swift]: Dont need `self` here.

File: MovieApp/CrewCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    private func commonInit() {
+        assignViews()
+        style()
+        layout()
+    }
+    
+    private func assignViews() {
```
Comment (IvanKulundzic) [inline MovieApp/CrewCollectionViewCell.swift]: You can use the usual three methods we use here, `createViews`, `styleViews` and `defineLayoutForViews`. You can use them to on all views, and cells are views basically. 

Try to organize your code like that. And then, you can call, something like `buildViews` method to call all three methods in the init.

File: MovieApp/CrewCollectionViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        paragraphStyle.lineHeightMultiple = 1.4
+        nameLabel.attributedText = NSMutableAttributedString(string: name, attributes: [NSAttributedString.Key.paragraphStyle: paragraphStyle])
+        roleLabel.attributedText = NSMutableAttributedString(string: role, attributes: [NSAttributedString.Key.paragraphStyle: paragraphStyle])
+    }
+    
+    private func style() {
+        nameLabel.font = UIFont.boldSystemFont(ofSize: 14)
+        
+        roleLabel.font = UIFont.systemFont(ofSize: 14)
+
```
Comment (IvanKulundzic) [inline MovieApp/CrewCollectionViewCell.swift]: Newlines.

File: MovieApp/CrewCollectionViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    public func setText(name: String, role: String) {
+        let paragraphStyle = NSMutableParagraphStyle()
+        paragraphStyle.lineHeightMultiple = 1.4
+        nameLabel.attributedText = NSMutableAttributedString(string: name, attributes: [NSAttributedString.Key.paragraphStyle: paragraphStyle])
+        roleLabel.attributedText = NSMutableAttributedString(string: role, attributes: [NSAttributedString.Key.paragraphStyle: paragraphStyle])
+    }
+    
+    private func style() {
+        nameLabel.font = UIFont.boldSystemFont(ofSize: 14)
+
```
Comment (IvanKulundzic) [inline MovieApp/CrewCollectionViewCell.swift]: Newline.

File: MovieApp/InputView.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        self.labelText = ""
+        self.inputFieldDefaultText = ""
+        super.init(coder: aDecoder)
+        self.commonInit()
+    }
+    
+    private func commonInit() {
```
Comment (IvanKulundzic) [inline MovieApp/InputView.swift:34]: Same as with the cell. Use `createViews`, `styleViews` and `defineLayoutForViews`.

File: MovieApp/SceneDelegate.swift [hunk 7]
```diff
@@ -12,11 +12,18 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
...
-    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
-        // Use this method to optionally configure and attach the UIWindow `window` to the provided UIWindowScene `scene`.
-        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
-        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+    func scene(
+        _ scene: UIScene,
+        willConnectTo session: UISceneSession,
+        options connectionOptions: UIScene.ConnectionOptions
+    ){
```
Comment (IvanKulundzic) [inline MovieApp/SceneDelegate.swift]: Add spacing after parentheses.

File: MovieApp/SceneDelegate.swift [hunk 8]
```diff
@@ -12,11 +12,18 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
...
+        guard let windowScene = scene as? UIWindowScene // 1
+        else{ return }
```
Comment (IvanKulundzic) [inline MovieApp/SceneDelegate.swift]: Spacing.

File: MovieApp/SceneDelegate.swift [hunk 9]
```diff
@@ -12,11 +12,18 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
...
+        
+        window = UIWindow(windowScene: windowScene)     // 2
```
Comment (IvanKulundzic) [inline MovieApp/SceneDelegate.swift]: What are these comments used for?

File: MovieApp/MovieBannerView.swift [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    
+    
+    init(details: MovieDetailsModel?) {
+        self.details = details
+        super.init(frame: .zero)
+        self.commonInit()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        self.details = nil
```
Comment (IvanKulundzic) [inline MovieApp/MovieBannerView.swift:32]: Why are details here set to nil?

File: MovieApp/MovieBannerView.swift [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+//
+//  Created by endava-bootcamp on 28.03.2023..
+//
+
+import UIKit
+import PureLayout
+import MovieAppData
+
+class MovieBannerView: UIView {
+    let scoreLabel = UILabel()
```
Comment (IvanKulundzic) [inline MovieApp/MovieBannerView.swift:13]: Group `var` and `let`.

File: MovieApp/MovieBannerView.swift [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    let dateLabel = UILabel()
+    let categories = UILabel()
+    let starButton = UIButton()
+    var backgroundImage = UIImageView()
+    var categoriesString = ""
+    
+    let starButtonSize = 32.0
+    
+    var details: MovieDetailsModel?
+
```
Comment (IvanKulundzic) [inline MovieApp/MovieBannerView.swift]: Extra newlines.

File: MovieApp/LogInViewController.swift [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    
+    init (){
+        super.init(nibName: nil, bundle: nil)
+        let color = UIColor(red: 11/255, green: 37/255, blue: 63/255, alpha: 1)
+        view.backgroundColor = color
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
Comment (IvanKulundzic) [inline MovieApp/LogInViewController.swift]: This part is not needed. You can set the view background color in a method. For example, in `styleViews`. No need to use an init for it.

File: MovieApp/LogInViewController.swift [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+        
+    }
+    
+    func createViews() {
+        signInLabel = UILabel()
+        view.addSubview(signInLabel)
+        signInLabel.text = "Sign in"
```
Comment (IvanKulundzic) [inline MovieApp/LogInViewController.swift]: This kind of customization should be in `styleViews`. Fix thorough.

File: MovieApp/LogInViewController.swift [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        passwordInput.autoAlignAxis(toSuperviewAxis: .vertical)
+        passwordInput.autoPinEdge(.top, to: .bottom, of: emailInput, withOffset: 24)
+        
+        signInButton.autoPinEdge(.top, to: .bottom, of: passwordInput, withOffset: 48)
+        signInButton.autoPinEdge(toSuperviewEdge: .leading, withInset: 32)
+        signInButton.autoPinEdge(toSuperviewEdge: .trailing, withInset: 32)
+        signInButton.autoSetDimension(.height, toSize: 40)
+    }
+    
+    override var preferredStatusBarStyle: UIStatusBarStyle {
```
Comment (IvanKulundzic) [inline MovieApp/LogInViewController.swift]: This should be on the top, after `var` and `let`.

File: MovieApp/MovieBannerView.swift [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+        self.addSubview(categories)
+        let cat = [MovieCategoryModel.action, MovieCategoryModel.adventure, MovieCategoryModel.drama, MovieCategoryModel.scienceFiction]
+        categoriesString = cat
+            .map {(category: MovieCategoryModel) -> String in return capitalSplitString(string: String(describing: category)).capitalized}
+            .joined(separator: ", ")
+        
+        self.addSubview(starButton)
+        self.addSubview(backgroundImage)
+        self.sendSubviewToBack(backgroundImage)
+
```
Comment (IvanKulundzic) [inline MovieApp/MovieBannerView.swift]: Newline.

File: MovieApp/MovieBannerView.swift [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+            dateLabel.text = dateFormatterNew.string(from: date!)
+        }
+        
+        self.addSubview(categories)
+        let cat = [MovieCategoryModel.action, MovieCategoryModel.adventure, MovieCategoryModel.drama, MovieCategoryModel.scienceFiction]
+        categoriesString = cat
+            .map {(category: MovieCategoryModel) -> String in return capitalSplitString(string: String(describing: category)).capitalized}
+            .joined(separator: ", ")
+        
+        self.addSubview(starButton)
```
Comment (IvanKulundzic) [inline MovieApp/MovieBannerView.swift]: Dont need `self` here.

File: MovieApp/LogInViewController.swift [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+//  LogInViewController.swift
+//  MovieApp
+//
+//  Created by endava-bootcamp on 24.03.2023..
+//
+import UIKit
+import PureLayout
+
+class LogInViewController: UIViewController {
+
```
Comment (IvanKulundzic) [inline MovieApp/LogInViewController.swift]: Newline.

File: MovieApp/LogInViewController.swift [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+
```
Comment (IvanKulundzic) [inline MovieApp/LogInViewController.swift]: Newline.

File: MovieApp/MovieDetailsViewController.swift [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+class MovieDetailsViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
+    var movieBannerView: MovieBannerView!
+    var details: MovieDetailsModel?
+    var overviewLabel: UILabel!
+    var summary: UILabel!
+    var roleCollectionView: UICollectionView!
+    var crewList: [MovieCrewMemberModel]!
+    let reuseIdentifier = "cell"
+
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift:20]: Newline.

File: MovieApp/MovieDetailsViewController.swift [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        details = MovieUseCase().getDetails(id: 111161)
+        guard let details else {return}
+        crewList = details.crewMembers
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift]: Newline.

File: MovieApp/MovieDetailsViewController.swift [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+    var roleCollectionView: UICollectionView!
+    var crewList: [MovieCrewMemberModel]!
+    let reuseIdentifier = "cell"
+    
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        details = MovieUseCase().getDetails(id: 111161)
+        guard let details else {return}
+        crewList = details.crewMembers
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift]: You can extract this part to a method. What happens when the details are nil?

File: MovieApp/MovieDetailsViewController.swift [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+        
+    }
+    
+    private func createViews() {
+        guard let details else {return}
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift]: You have this guard already? Also, spacing after curly brackets.

File: MovieApp/MovieDetailsViewController.swift [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+        
+        roleCollectionView.autoPinEdge(.top, to: .bottom, of: summary, withOffset: 27)
+        roleCollectionView.autoPinEdge(toSuperviewEdge: .leading, withInset: 16)
+        roleCollectionView.autoPinEdge(toSuperviewEdge: .trailing, withInset: 16)
+        roleCollectionView.autoPinEdge(toSuperviewEdge: .bottom)
+    }
+    
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return crewList.count
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift]: You can omit the return keyword when its a one-liner.

File: MovieApp/MovieDetailsViewController.swift [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+        let cell = roleCollectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as! CrewCollectionViewCell
+        
+        let crewMember = crewList[indexPath.item]
+        
+        cell.setText(name: crewMember.name, role: crewMember.role)
+        
+        return cell
+    }
+    
+    override var preferredStatusBarStyle: UIStatusBarStyle {
```
Comment (IvanKulundzic) [inline MovieApp/MovieDetailsViewController.swift]: This should be at the beginning of the file, after the `var` and `let`.

**Review by IvanKulundzic — CHANGES_REQUESTED** [review]

File: MovieApp/AppDelegate.swift (hunk 1 above)
Comment (JuraH32) [reply MovieApp/AppDelegate.swift]: Not sure, removed it now

File: MovieApp/MovieBannerView.swift (hunk 10 above)
Comment (JuraH32) [reply MovieApp/MovieBannerView.swift:32]: Didn't know what should be the default value when using this init which I don't think is ever going to be used.

File: MovieApp/MovieDetailsViewController.swift (hunk 22 above)
Comment (JuraH32) [reply MovieApp/MovieDetailsViewController.swift]: Currently, it doesn't work if the details are nil, so it will be changed when the details are dynamically fetched and not hardcoded to be fetched from a single id.

**Review by IvanKulundzic — APPROVED** [review]
